### PR TITLE
Fix Item ID variance between seeds and generation issues on Linux

### PR DESCRIPTION
**Pull request by dclaganiere**
- Added an item table with constant item IDs
- Updated item groups with new groups to categorize all items
- Fixed usage of absolute path when loading resources with `pkgutil`

### CODE REVIEW

File: worlds/rabi_ribi/__init__.py [hunk 1]
```diff
@@ -38,42 +38,17 @@ class RabiRibiWorld(World):
...
 
-    item_name_to_id: Dict[str, int] = {
-        name: id_num for
-        id_num, name in enumerate(item_set, base_id)
-    }
-    item_name_to_id["Nothing"] = base_id + len(item_name_to_id)
+    item_name_to_id = {name: data.code for name, data in item_table.items()}
     location_name_to_id: Dict[str, int] = {
         name: id_num for
         id_num, name in enumerate(get_all_possible_locations(), base_id)
```
Comment (tdkollins) [inline worlds/rabi_ribi/__init__.py:49]: You don't need to do so on this PR, but heads up the locations are going to have the same problem if anyone ever makes the locations variable.

File: worlds/rabi_ribi/items.py [hunk 2]
```diff
@@ -1,41 +1,139 @@
 """This module represents item definitions for Rabi-Ribi"""
-from typing import List
+from typing import Dict, List, NamedTuple, Optional, Set
 
-from BaseClasses import Item
+from BaseClasses import Item, ItemClassification
 from .existing_randomizer.visualizer import load_item_locs
 
+rabi_ribi_base_id: int = 8350438193300
```
Comment (tdkollins) [inline worlds/rabi_ribi/items.py:7]: Might make more sense to stick this ID in `utility.py` (with a `get_rabi_ribi_base_id()`). This ID is used for locations as well as items.

**Comment by tdkollins** [comment]
Thanks for the changes here! I tested generation on Windows to confirm the path changes didn't break anything on that end too 👍

**Review by tdkollins — APPROVED** [review]
A couple Nits but overall looks good. Thanks!

File: worlds/rabi_ribi/__init__.py (hunk 1 above)
Comment (dclaganiere) [reply worlds/rabi_ribi/__init__.py]: Makes sense; I created some new tables in `locations.py` for all the locations, including the post game and gift items for the future. I also added a names module for item and location names.

File: worlds/rabi_ribi/locations.py [hunk 3]
```diff
@@ -1,324 +1,393 @@
...
+}
+
+# Shufflable Gift Item Locations
+shufflable_gift_item_locations = {
+    LocationName.p_hairpin                    : get_rabi_ribi_base_id() + 0x98,
+    LocationName.speed_boost                  : get_rabi_ribi_base_id() + 0x99,
+    LocationName.bunny_strike                 : get_rabi_ribi_base_id() + 0x9A,
+}
+
+# Egg Locations
```
Comment (tdkollins) [inline worlds/rabi_ribi/locations.py:200]: Why separate the egg locations from the other item locations in the area?

File: worlds/rabi_ribi/logic_helpers.py [hunk 4]
```diff
@@ -380,25 +375,25 @@ def convert_existing_rando_rule_to_ap_rule(existing_rule: object, player: int, r
...
-            "Tm Nieve": lambda state: state.has("Nieve Recruit", player),
-            "Tm Nixie": lambda state: state.has("Nixie Recruit", player),
-            "Tm Aruraune": lambda state: state.has("Aruraune Recruit", player),
-            "Tm Seana": lambda state: state.has("Seana Recruit", player),
-            "Tm Lilith": lambda state: state.has("Lilith Recruit", player),
-            "Tm Vanilla": lambda state: state.has("Vanilla Recruit", player),
-            "Tm Chocolate": lambda state: state.has("Chocolate Recruit", player),
-            "Tm Kotri": lambda state: state.has("Kotri Recruit", player),
-            "Tm Keke Bunny": lambda state: state.has("Keke Bunny Recruit", player),
+            "Piko Hammer Leveled": lambda state: state.has(ItemName.piko_hammer, player),
```
Comment (tdkollins) [inline worlds/rabi_ribi/logic_helpers.py:378]: Side note, if you ever update the logic, you might want to take a look at OpBacktrack (mentioned in the documentation here)

File: worlds/rabi_ribi/regions.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+
+        speed_boost = RabiRibiLocation(self.player, ItemName.speed_boost, None, regions[LocationName.town_main])
+        speed_boost.place_locked_item(RabiRibiItem(ItemName.speed_boost, ItemClassification.progression, None, self.player))
+        regions[LocationName.town_main].locations.append(speed_boost)
+
+    def _get_region_name_list(self):
+        return [
+            convert_existing_rando_name_to_ap_name(name) for \
+            name in self.randomizer_data.graph_vertices
+        ]
```
Comment (tdkollins) [inline worlds/rabi_ribi/regions.py:305]: Didn't look at this super closely... Is there anything new here or has this just changed to a dedicated region file?

File: worlds/rabi_ribi/utility.py [hunk 6]
```diff
@@ -1,14 +1,12 @@
...
 
-def get_world_directory():
-    """
-    Get the base rabi_ribi world directory
-    """
-    return os.path.dirname(os.path.abspath(__file__))
+rabi_ribi_base_id: int = 8350438193300
+
+def get_rabi_ribi_base_id() -> int:
+    return rabi_ribi_base_id
```
Comment (tdkollins) [inline worlds/rabi_ribi/utility.py:9]: 💯

**Comment by tdkollins** [comment]
Looks good overall, left a couple questions. Do note I did skim over a lot of the changes from item/location strings into attributes and didn't confirm the behavior super closely. Thanks here again!

File: worlds/rabi_ribi/logic_helpers.py (hunk 4 above)
Comment (dclaganiere) [reply worlds/rabi_ribi/logic_helpers.py:378]: Yeah, I noticed the commented out prereqs when looking at the constraints, but I haven't thought too much about it yet.

File: worlds/rabi_ribi/locations.py (hunk 3 above)
Comment (dclaganiere) [reply worlds/rabi_ribi/locations.py:200]: Honestly, it's not necessary for now, if ever. I was simply copying the items in order from `location_items.txt` , which separated them.

I guess it does mean we could do something like exclude easter egg locations if we ever wanted to. :)

Comment (tdkollins) [reply worlds/rabi_ribi/locations.py:200]: I forgot `locations_items.txt` did that. Makes sense 👍

**Pull request merged**